utils/controller/controller.go

```diff
@@ -18,6 +23,43 @@ import (
 	logutil "github.com/argoproj/argo-rollouts/utils/log"
 )
 
+// processNextWatchObj will process a single object from the watch by seeing if
+// that object is in an index and enqueueing the value object from the indexer
+func processNextWatchObj(watchEvent watch.Event, queue workqueue.RateLimitingInterface, indexer cache.Indexer, index string) {
+	obj := watchEvent.Object
+	acc, err := meta.Accessor(obj)
+	if err != nil {
+		log.Errorf("Error processing object from watch: %v", err)
+		return
+	}
+	objsToEnqueue, err := indexer.ByIndex(index, fmt.Sprintf("%s/%s", acc.GetNamespace(), acc.GetName()))
+	if err != nil {
+		log.Errorf("Cannot process indexer: %s", err.Error())
+		return
+	}
+	for i := range objsToEnqueue {
+		Enqueue(objsToEnqueue[i], queue)
+	}
+}
+
+// WatchResource is a long-running function that will continually call the
+// processNextWatchObj function in order to watch changes on a resource kind
+// and enqueue a different resources kind that interact with them
+func WatchResource(client dynamic.Interface, gvk schema.GroupVersionResource, queue workqueue.RateLimitingInterface, indexer cache.Indexer, index string) {
+	log.Infof("Starting watch on resource '%s'", gvk.Resource)
+	watch, err := client.Resource(gvk).Watch(metav1.ListOptions{})
+	if k8serrors.IsNotFound(err) {
+		log.Infof("Watch for resource '%s' failed: resource not found", gvk.Resource)
+		return
+	}
+	if err != nil {
+		panic(err)
+	}
+	for watchEvent := range watch.ResultChan() {
+		processNextWatchObj(watchEvent, queue, indexer, index)
+	}
+}
```

Comment:
The problem with this function, is that it returns indiscriminately if the resource GVK doesn't exist,  vs. normal failures that happen during the watch. The difference is that when the GVK doesn't exist, we want to backoff longer (e.g. 5 minutes). However if the watch fails simply because the stream got closed, we probably want to retry immediately. Otherwise, we will incur a 1 minute delay every time the watch is closed (which can actually be frequent).